src/Common/LRUFileCache.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+#include <boost/functional/hash.hpp>
+#include <boost/noncopyable.hpp>
+#include <map>
+
+#include <Common/logger_useful.h>
+#include <Common/FileSegment.h>
+#include <Common/IFileCache.h>
+
+
+namespace DB
+{
+
+class LRUFileCache final : public IFileCache
+{
+public:
+    LRUFileCache(
+        const String & cache_base_path_,
+        const FileCacheSettings & cache_settings_);
+
+    FileSegmentsHolder getOrSet(const Key & key, size_t offset, size_t size, bool is_persistent) override;
+
+    FileSegmentsHolder get(const Key & key, size_t offset, size_t size) override;
+
+    FileSegments getSnapshot() const override;
+
+    void initialize() override;
+
+    void removeIfExists(const Key & key) override;
+
+    void removeIfReleasable(bool remove_persistent_files) override;
+
+    std::vector<String> tryGetCachePaths(const Key & key) override;
```

Comment:
It has `try` prefix because it can return empty because key doesn't exist?

Reply:
yes, empty `std::vector<String>`